gnovm/pkg/gnolang/preprocess.go

```diff
@@ -3779,16 +3791,486 @@ func assertTypeDeclNoCycle2(store Store, last BlockNode, x Expr, stack *[]Name,
+
+			if un != "" {
+				return un
+			}
+		}
+	case *SwitchClauseStmt:
+		for _, rh := range s.Cases {
+			un := findUndefined2SkipLocals(store, last, rh, t)
+
+			if un != "" {
+				return un
+			}
+		}
+
+		for _, b := range s.Body {
+			un := findUndefinedStmt(store, last, b, t)
+
+			if un != "" {
+				return un
+			}
+		}
+
+	case *ExprStmt:
+		return findUndefined2SkipLocals(store, last, s.X, t)
+	case *AssignStmt:
+		for _, rh := range s.Rhs {
+			un := findUndefined2SkipLocals(store, last, rh, t)
+
+			if un != "" {
+				return un
+			}
+		}
+	case *IfStmt:
+		un := findUndefinedStmt(store, last, s.Init, t)
+		if un != "" {
+			return un
+		}
+
+		un = findUndefined2SkipLocals(store, last, s.Cond, t)
+		if un != "" {
+			return un
+		}
+
+		un = findUndefinedStmt(store, last, &s.Else, t)
+		if un != "" {
+			return un
+		}
+
+		un = findUndefinedStmt(store, last, &s.Then, t)
+		if un != "" {
+			return un
+		}
+	case *IfCaseStmt:
+		for _, b := range s.Body {
+			un := findUndefinedStmt(store, last, b, t)
+
+			if un != "" {
+				return un
+			}
+		}
+	case *ReturnStmt:
+		for _, b := range s.Results {
+			un := findUndefined2SkipLocals(store, last, b, t)
+			if un != "" {
+				return un
+			}
+		}
+	case *RangeStmt:
+		un := findUndefined2SkipLocals(store, last, s.X, t)
+		if un != "" {
+			return un
+		}
+
+		for _, b := range s.Body {
+			un := findUndefinedStmt(store, last, b, t)
+			if un != "" {
+				return un
+			}
+		}
+	case *ForStmt:
+		un := findUndefinedStmt(store, s, s.Init, t)
+		if un != "" {
+			return un
+		}
+
+		un = findUndefined2SkipLocals(store, s, s.Cond, t)
+		if un != "" {
+			return un
+		}
+
+		un = findUndefinedStmt(store, s, s.Post, t)
+		if un != "" {
+			return un
+		}
+
+		for _, b := range s.Body {
+			un := findUndefinedStmt(store, last, b, t)
+			if un != "" {
+				return un
+			}
+		}
+	case *BranchStmt:
+	case nil:
+		return ""
+	default:
+		panic(fmt.Sprintf("findUndefinedStmt: %T not supported", s))
+	}
+	return ""
+}
+
+func getGlobalValueRef(sb BlockNode, store Store, n Name) *TypedValue {
+	sbb := sb.GetStaticBlock()
+	idx, ok := sb.GetLocalIndex(n)
+	bb := &sb.GetStaticBlock().Block
+	bp := sb.GetParentNode(store)
+
+	for {
+		if ok && sbb.Types[idx] != nil && (bp == nil || bp.GetParentNode(store) == nil) {
+			return bb.GetPointerToInt(store, int(idx)).TV
+		} else if bp != nil {
+			idx, ok = bp.GetLocalIndex(n)
+			sbb = bp.GetStaticBlock()
+			bb = sbb.GetBlock()
+			bp = bp.GetParentNode(store)
+		} else {
+			return nil
+		}
+	}
+}
+
+func findUndefinedGlobal(store Store, last BlockNode, x Expr, t Type) (un Name) {
+	if x == nil {
+		return
+	}
+	switch cx := x.(type) {
+	case *NameExpr:
+		if tv := getGlobalValueRef(last, store, cx.Name); tv != nil {
+			return
+		}
+
+		if _, ok := UverseNode().GetLocalIndex(cx.Name); ok {
+			// XXX NOTE even if the name is shadowed by a file
+			// level declaration, it is fine to return here as it
+			// will be predefined later.
+			return
+		}
```

Comment:
It looks like `findUndefinedGlobal` and `findUndefined2` share most of their logic, with the main difference being the search space. can we potentially modify `findUndefined2` slightly to handle the different search spaces and reuse the common logic?